### DISCUSSION TITLE
Queries with something else than ?

### DISCUSSION THREAD

**Discussion by daleif** [mojolicious/mojo#1691] [Ideas]
I've been playing a bit with an application where we load a PDF in an iframe on the left side of the screen. The build in PDF viewers in Firefox and Chrome can react to parts of the URL you give the PDF URL, but you need `#` not `?` as the separator before the "query". Do we have any way of controlling the separator in say `->query(...)`.

In my POC I just  used

```
 my $url = url_for('pdf'})->query(page => 2);
```

and did a replacement on `$url`, might have been nice with something cleaner.

As an example `#page=2&zoom=150` would start at page 2 with a specific zoom level.

**Comment by kiwiroy** [comment]
Is that a query string or a fragment composed like a query_string?

    $ perl -Mojo -E 'my $pdf = l q{/output.pdf}; say $pdf->fragment(q{page=2&zoom=150})'
    /output.pdf#page=2&zoom=150

**Reply by daleif** [reply]
Ahh, that is it, never even seen that one before.

Thanks

**Comment by jberger** [comment]
You could even still use a new instance of Mojo::Parameters to build the query-like content for the fragment

**Reply by daleif** [reply]
yes, that makes it even cleaner. Especially when we add more config options for zoom etc.

**Comment by jhthorsen** [comment]
Sounds like you should use query params and not a fragment for this though. Have you considered using https://developer.mozilla.org/en-US/docs/Web/API/History/pushState on the client side?

**Reply by daleif** [reply]
It works just fine no javascript needed. Though I might switch to using only `pdf.js` as the internal pdf viewer in chrome does not quite support the same query options as the one in Firefox does (they use `pdf.js`).

<https://github.com/mozilla/pdf.js/wiki/Viewer-options>

For now I'm using 

```
 url_for('pdf')->fragment(Mojo::Parameters->new(page => 2, zoom => 'page-width', pagemode => 'none' ));
```

which works just fine for the proof of concept I was attempting.
